promql/engine.go

```diff
@@ -925,6 +939,7 @@ func (ev *evaluator) rangeEval(f func([]parser.Value, *EvalNodeHelper) Vector, e
 							// past points at the next step.
 							matrixes[i][si].Points = series.Points[1:]
 							ev.currentSamples++
+							ev.updateSamples(ev.currentSamples - tempNumSamples)
```

Comment:
For calculating the peak shouldn't we always be passing in currentSamples, as that's what's in memory?

Reply:
Well, the currentSamples count is already being maintained inside the `updateSamples` func, and hence passing currentSamples would be like `currentSamples * 2 + tempNumSamples`. Hence, did this way.

Reply:
This seems like you're more than double counting then, as you descend down the tree you'll be counting the already-computed LHS of parents as being processed again.

Reply:
Well not sure on this, but each time we increase the currentSamples, it refers to the samples being increased in down the tree. that means, we need to save them in totalSamples as well and hence totalSamples are increased by the amount of addition to the currentSamples only.

Also, we need to capture the total samples involved in the evaluation of expr like binary which will involve combination of samples from more than one `currentSamples` tree. 

So in short, adding the total count of samples by the amount of increment done to the currentSamples should do the thing.

Reply:
@brian-brazil any views?

Reply:
I'm not sure, and even if that were true the code seems fragile.

Reply:
Sure, will try to have a second go through the code and try to dig more in a reliable way if possible.

Reply:
I think, that should be `ev.updateSamples(1)`. What do you say? @brian-brazil

Reply:
Well, I went through this and realized that we need two inputs to the `updateSamples()`, one for the total samples, and the other for having instantaneous peak samples in order to capture the peak samples. I don't think that now, samples are being caught more than once, however still give a look at the implementation.